From d6e628a42ed2b2dbec719473ff2434475910adfb Mon Sep 17 00:00:00 2001
From: Xi Ruoyao <xry111@xry111.site>
Date: Sun, 28 Dec 2025 18:18:19 +0800
Subject: [PATCH 45/63] loonggpu: Handle the error from gsgpu_driver_load_kms
 correctly

With a 4KiB-page kernel, gsgpu_gart_init will fail because it requires
PAGE_SIZE >= GSGPU_GPU_PAGE_SIZE.  The error will then be propagated by
mmu_gart_init, mmu_sw_init, gsgpu_device_ip_init (calling mmu_sw_init
via a sw_init function pointer), gsgpu_device_init, and
gsgpu_driver_load_kms.  But then loongson_vga_pci_register does not
check the return value of gsgpu_driver_load_kms, leading to kernel oops
and hangs.

Handle the error gracefully to allow the 4KiB-page kernel continuing to
function (despite without loonggpu support).

Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 loonggpu/loonggpu_drv.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/loonggpu/loonggpu_drv.c b/loonggpu/loonggpu_drv.c
index 5979dc0..ce78f53 100644
--- a/loonggpu/loonggpu_drv.c
+++ b/loonggpu/loonggpu_drv.c
@@ -651,6 +651,8 @@ static int loongson_vga_pci_register(struct pci_dev *pdev,
 		pci_set_drvdata(loongson_gpu_pdev, dev);
 
 	ret = loonggpu_driver_load_kms(dev, pci_gpu_flags);
+	if (ret)
+		goto err_pci;
 
 retry_init:
 	ret = drm_dev_register(dev, pci_gpu_flags);
-- 
2.52.0

